Extend Swap Contract with a new behavior to take a fixed fee

piotrwitek opened the pull request:
This pull request introduces a new `FeeType` enum to the Swap contract, introducing ability to take a fee in the fixed amount alongside existing percentage fee.

Additionally implemented couple of chores, like enhanced error handling by providing clear messages for missing environment variables. the Hardhat Tenderly setup, and various tests have been improved to ensure better reliability. Documentation has been added to support these changes.

piotrwitek commented on packages/dma-contracts/contracts/swap/Swap.sol:
```diff
@@ -127,13 +128,41 @@ contract Swap {
+  ) internal returns (uint256 amount) {
+    amount = fromAmount.sub(fee);
+
+    if (fee > 0) {
```
@zerotucks should I also add a check somewhere here to guard when the fee is greater than fromAmount?

piotrwitek replied:
Done, if fee is greater than fromAmount then we'll take whatever is available as fee. Confirmed by Chris.

robercano reviewed (approved):
Looks great!! I've mainly reviewed the solidity changes and the look awesome

piotrwitek committed:
workflow updated

piotrwitek committed:
Added new swap contract integration and fixed fee utilities
Also refactored fee resolver to make decision logic which fee to use based on asset correlation

piotrwitek committed:
Extend Swap contract with a new enum FeeType (Percentage/Fixed values) to allow choosing the fee type calculation for the fee taken. Also added new calculation handler for fixed fee type.

piotrwitek committed:
Removed duplicated SlippageSaved event emission

piotrwitek committed:
Refactor deployment scripts to use updated import paths and enhance README with additional npm scripts

Updated existing tests and added new tests covering implemented functionalities

piotrwitek committed:
updated conf

piotrwitek committed:
Merge commit '9a291658e48a28d4b08fc057445998183cb9b47e' into pw/Extend-Swap-contract-data-with-new-enum-FeeType

piotrwitek committed:
Integrated new fee resolver logic in Ajna and Morpho protocol implementation

piotrwitek committed:
Fixed type issue

piotrwitek committed:
Merge commit 'c7483129b39e4e63ff26b06994ed9a4d36bdcba3' into pw/Extend-Swap-contract-data-with-new-enum-FeeType

piotrwitek committed:
Fixed merge conflicts

piotrwitek committed:
fixed script

piotrwitek committed:
bumped library version

piotrwitek committed:
linted readme

piotrwitek committed:
Updated to use new swap_3 contract address and updated deployment scripts

piotrwitek committed:
Refactor OneInch API endpoint variable names and update swap URL formatting

piotrwitek committed:
Add Morpho subgraph queries and client integration

piotrwitek committed:
Fixed fee calculations

piotrwitek committed:
Add error handling for missing protocol ID in fixed fee calculation

piotrwitek committed:
Fix casing for SPARK protocol ID in fee service utilities

piotrwitek committed:
Refactor fee calculation methods to use ResolvedFee type and improve clarity in swap fee handling

piotrwitek committed:
Add feeType parameter to swapData calldata type for enhanced fee handling

piotrwitek committed:
Enabled contract verification and tenderly deployment setup script

piotrwitek committed:
Refactor Tenderly deployment initialization to accept provider as a parameter

piotrwitek committed:
Enable deployment for Swap and SwapAction configurations in mainnet settings

piotrwitek committed:
Enable logging of sent txErc20 in tenderlyDeployInit script

piotrwitek committed:
Added working tendrly verification

piotrwitek committed:
Add fee tier addition to new Swap contract and enhance Tenderly deployment initialization

piotrwitek committed:
fixed build issue

piotrwitek committed:
Fix build issue with packages dependency

piotrwitek committed:
Added debugging information

piotrwitek committed:
Update Tenderly deployment initialization with new balance values

piotrwitek committed:
better debugging

piotrwitek committed:
Add debug log for zero fixed fee when opening an earn position

piotrwitek committed:
v0.7.0-rc.1

piotrwitek committed:
Merge commit 'ab3f190eecc16fc8da486ec53451009d61f2976e' into pw/Extend-Swap-contract-data-with-new-enum-FeeType

piotrwitek committed:
v0.7.0-rc.2

piotrwitek committed:
Fixed missing morpho marketid

piotrwitek committed:
Merge commit '0d2a150edd7baf6884cefc6d0bea3cd589694345' into pw/Extend-Swap-contract-data-with-new-enum-FeeType

piotrwitek committed:
fix import after merge

piotrwitek committed:
Fix type issue

piotrwitek committed:
Fixed dependencies and lock file after merge

piotrwitek committed:
v0.7.0-rc3

piotrwitek committed:
Fixed issues with timestamp calculations and extended tests

piotrwitek committed:
Refactor fee as big number to fix decimals issue

piotrwitek committed:
Fixed issues with close fee calculations

piotrwitek committed:
Add feeType to WithFee type and update related calculations

piotrwitek committed:
Refactor fee calculation logic to handle event ordering and improve logging

piotrwitek committed:
Refactor fee calculation to improve clarity and logging for event periods

piotrwitek committed:
Add feeType handling to collateral swap calculations for improved fee management

piotrwitek committed:
Add feeType to position test fees for improved fee handling

piotrwitek committed:
Merge commit '57668dbd7888d3c648f7c31f39bab48b03e49d16' into pw/Extend-Swap-contract-data-with-new-enum-FeeType

piotrwitek committed:
v0.7.0-rc4

piotrwitek committed:
up tests

piotrwitek committed:
Fixed parameter type in test

piotrwitek committed:
fixed fee param in tests

piotrwitek committed:
Refactor fee logging and improve error messages in fee resolver

piotrwitek committed:
Enhance fixed fee resolver with additional parameter for increasing risk

piotrwitek committed:
v0.7.0-rc5

piotrwitek committed:
Refactor fee calculation logic to apply the fee limit - cannot be higher than 1 year of the current debt

piotrwitek committed:
Add debt property to Ajna position response

piotrwitek committed:
v0.7.0-rc6

piotrwitek committed:
prep deployment
